config/opal_config_3rdparty.m4

```diff
@@ -13,44 +13,58 @@ dnl
+dnl not defined, then opal_pkg_mode will not be internal.  If
+dnl "disabled ok" is not defined, then opal_pkg_mode will not be
+dnl "disabled".
 dnl
 dnl If m4_ifdef(internal support) does not evaluate to true (ie, at
 dnl autogen time), the references to internal in the help strings will
 dnl be removed and internal will not be a supported option.
 dnl
+dnl If m4_ifval(ddisbaled ok) does not evaluete to true (ie, at autogen
+dnl time), then --without-pkg will not be a valid configure option and
+dnl will raise an error.
+dnl
 dnl $1: short package name
 dnl $2: long pacakage name
 AC_DEFUN([OPAL_3RDPARTY_WITH], [
-    m4_ifdef([$3],
-        [AC_ARG_WITH([$1],
-            [AS_HELP_STRING([--with-$1(=DIR)],
-                           [Build $2 support.  DIR can take one of three values: "internal", "external", or a valid directory name.  "internal" forces Open MPI to use its internal copy of $2.  "external" forces Open MPI to use an external installation of $2.  Supplying a valid directory name also forces Open MPI to use an external installation of $2, and adds DIR/include, DIR/lib, and DIR/lib64 to the search path for headers and libraries. Note that Open MPI no longer supports --without-$1.  If no argument is specified, Open MPI will search default locations for $2 and fall back to an internal version if one is not found.])])
-
-         AC_ARG_WITH([$1-libdir],
-            [AS_HELP_STRING([--with-$1-libdir=DIR],
-             [Search for $2 libraries in DIR.  Should only be used if an external copy of $2 is being used.])])],
-        [AC_ARG_WITH([$1],
-            [AS_HELP_STRING([--with-$1(=DIR)],
-                           [Build $2 support.  DIR can take one of two values:  "external" or a valid directory name.  "external" forces Open MPI to use an external installation of $2.  Supplying a valid directory name also forces Open MPI to use an external installation of $2, and adds DIR/include, DIR/lib, and DIR/lib64 to the search path for headers and libraries. Note that Open MPI no longer supports --without-$1.  If no argument is specified, Open MPI will search default locations for $2 and error if one is not found.])])
+    m4_ifval([$4],
+        [m4_ifdef([$3],
+            [AC_ARG_WITH([$1],
+                [AS_HELP_STRING([--with-$1(=DIR)],
+                                [Build $2 support.  DIR can take one of four values: "internal", "external", "no", or a valid directory name.  "internal" forces Open MPI to use its internal copy of $2.  "external" forces Open MPI to use an external installation of $2.  Supplying a valid directory name also forces Open MPI to use an external installation of $2, and adds DIR/include, DIR/lib, and DIR/lib64 to the search path for headers and libraries. "no" means that Open MPI will not build components that require this package.  If no argument is specified, Open MPI will search default locations for $2 and fall back to an internal version if one is not found.])])],
```

Comment:
I don't quite understand why this is needed three times with slightly wording. Can you elaborate?

Reply:
It's actually 4 times :).  There are four cases it covers:

1. The package has an internal copy and cannot be disabled
2. The package does not have an internal copy and cannot be disabled
3. The package has an internal copy and can be disabled
4. The package does not have an internal copy and can be disabled

I tried to be clever about composing the strings, but AC_ARG_WITH does some clever things with the first and second argument to make `./configure --help` work (it's actually a giant pre-defined string, not generated at configure time) and I couldn't get our clevernesses to work together.

As for why those four options, libevent, hwloc, and PMIx cannot be disabled.  We have to find a usable package for those three packages.  PRRTE can be disabled; you can build an OPen MPI build without PRRTE.  Any of the four packages normally are build in the release tarball, but autogen.pl allows that not to be the case, so any one of the four can be missing an "internal" option.